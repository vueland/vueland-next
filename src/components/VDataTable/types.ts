export interface IDataTableHeaderOptions {
  color?: string
  contentColor?: string
}

export interface IFooterPaginationOptions {
  buttonsColor?: string,
  displayColor?: string
}

export interface IDataTableFooterOptions {
  pagination?: IFooterPaginationOptions,
  color?: string
  contentColor?: string,
  rowsPerPageOptions?: Array<number>
  rowsPerPageText?: string
}
