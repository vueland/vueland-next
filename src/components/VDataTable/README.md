# VDataTable
```vue
  <v-data-table
    :cols="cols"
    :rows="rows"
    numbered
    checkbox
    @filter="testFunc"
    @checked="testFunc"
    @last-page="fetchItems"
  >
    <template v-slot:address="{ row, format }">
      <v-icon
        icon="fas fa-envelope"
        size="12"
        :color="format(row).includes('h') ? 'red':'blue'"
      />
      <span style="margin-left: 15px">{{ format(row) }}</span>
    </template>
    <template v-slot:toolbar>
      <v-button
        text="add"
        color="blue"
        elevation="3"
        @click="onClick"
      />
    </template>
  </v-data-table>
```
## Props

<table>
    <thead>
        <tr>
            <th colspan="3">Prop</th>
            <th>Type</th>
            <th>Default value</th>
            <th>Required</th>
            <th colspan="4">Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td colspan="3">rows</td>
            <td>Array</td>
            <td>null</td>
            <td>true</td>
            <td>array of objects which is represent table rows</td>
        </tr>
        <tr>
            <td colspan="3">cols</td>
            <td>Array</td>
            <td>null</td>
            <td>true</td>
            <td>array of objects which is represent table columns</td>
        </tr>
        <tr>
            <td colspan="3">rows-on-table</td>
            <td>Array</td>
            <td>[10, 15, 20, 25]</td>
            <td>false</td>
            <td>array of numbers for displaying the <br/> number of rows to display panel</td>
        </tr>
        <tr>
            <td colspan="3">color</td>
            <td>String</td>
            <td>white</td>
            <td>false</td>
            <td>table background color</td>
        </tr>
        <tr>
            <td colspan="3">header-color</td>
            <td>String</td>
            <td>white</td>
            <td>false</td>
            <td>table header background color</td>
        </tr> 
        <tr>
            <td colspan="3">align</td>
            <td>String</td>
            <td>left</td>
            <td>false</td>
            <td>aligns the display of table content.<br/> You can choose: left, center, right</td>
        </tr>
        <tr>
            <td colspan="3">dark</td>
            <td>Boolean</td>
            <td>false</td>
            <td>false</td>
            <td>if set true, table content will be<br/> displayed in white</td>
        </tr>
         <tr>
            <td colspan="3">state-out</td>
            <td>Boolean</td>
            <td>false</td>
            <td>false</td>
            <td>if set true, the filter event appears in the table,<br/> which can be handled by custom methods,<br/> respectively, internal filtering in the<br/> table will be disabled</td>
        </tr>
        <tr>
            <td colspan="3">numbered</td>
            <td>Boolean</td>
            <td>false</td>
            <td>false</td>
            <td>if set true, table rows will be numbered</td>
        </tr>
        <tr>
            <td colspan="3">checkbox</td>
            <td>Boolean</td>
            <td>false</td>
            <td>false</td>
            <td>if set true, table rows will <br/> be supplemented with a checkbox</td>
        </tr>
    </tbody>
</table>


## Events
<table>
    <thead>
        <tr>
            <th>Event name</th>
            <th>Value type</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>filter</td>
            <td>Object</td>
            <td>Object with "value" and "col" props</td>
        </tr>
        <tr>
            <td>checked</td>
            <td>Array</td>
            <td>Array of checked rows</td>
        </tr>
        <tr>
            <td>last-page</td>
            <td>Number</td>
            <td>The event will be triggered when the<br/> user reaches the last page, which makes <br/> it possible to load the necessary data asynchronously.<br/>As an argument, the handling function<br/> will get the index of the last<br/> current element</td>
        </tr>
    </tbody>
</table>

## Column object
```ts
        const column = {
          key: string,
          title: string,
          align?: 'left'|'center'|'right',
          resizeable?: boolean,
          sortable?: boolean,
          filterable?: boolean,
          cellClass?: string,
          filterClass?: string,
          rowCellClass?: string,
          show?: boolean,
          format?: (arg: any) => string
        }
```
<table>
    <thead>
        <tr>
            <th>Prop name</th>
            <th>Type</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>key</td>
            <td>String</td>
            <td>The property of the row object that<br/> will be displayed in this column</td>
        </tr>
        <tr>
            <td>title</td>
            <td>String</td>
            <td>Column title which will be displayed</td>
        </tr>
        <tr>
            <td>resizable</td>
            <td>Boolean</td>
            <td>If set true, will be able to change<br/> the size of the column </td>
        </tr>
         <tr>
            <td>sortable</td>
            <td>Boolean</td>
            <td>If set true, will be able to sort by<br/> current column</td>
        </tr>
        <tr>
            <td>filterable</td>
            <td>Boolean</td>
            <td>If set true, will be able to filter<br/> column by value</td>
        </tr>       
         <tr>
            <td>cellClass</td>
            <td>String</td>
            <td>custom class name for header cell<br/> column by value</td>
        </tr>         
        <tr>
            <td>filterClass</td>
            <td>String</td>
            <td>custom class name for header filter element<br/> column by value</td>
        </tr>        
        <tr>
            <td>rowCellClass</td>
            <td>String</td>
            <td>custom class name for table body column cells<br/> column by value</td>
        </tr>
        <tr>
            <td>show</td>
            <td>Boolean</td>
            <td>If set false, the column will be hidden.<br/> By default the column always is in<br/> visible state</td>
        </tr>
        <tr>
            <td>format</td>
            <td>Function</td>
            <td>You can handle the display of a row<br/> value in a column by using a function<br/> that takes a row object as an argument.<br/> If the function is set it will be used<br/> when sorting and filtering the row data.</td>
        </tr>
    </tbody>
</table>


