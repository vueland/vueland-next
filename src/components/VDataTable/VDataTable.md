# VDataTable
```vue
  <v-data-table
    :cols="cols"
    :rows="rows"
    toolbar
    numbered
    checkbox
    @filter="testFunc"
    @checked="testFunc"
    @last-page="fetchItems"
  />
```
## Props
  - rows
  - cols
  - rowCounts
  - color
  - headerColor
  - dark
  - stateOut
  - numbered
  - checkbox
  - align

## Events
  - filter
  - checked
  - lastPage


