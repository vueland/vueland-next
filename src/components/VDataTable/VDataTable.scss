@import "../../styles/scss/modules/global";
@import "variables";

.v-data-table {
  @include flex(flex, flex-start, flex-start, column);
  height: 100%;
  max-height: calc(100vh - 50px);
  border-radius: $table-border-radius;
  overflow: hidden;

  &__inner {
    width: 100%;
    padding: 0 2px 2px 2px;
    overflow: auto;
  }
}
