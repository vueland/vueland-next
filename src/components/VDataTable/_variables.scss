$table-row-min-width: calc(100% - 10px);
$table-row-height: 36px;
$table-header-height: 50px;
$table-footer-height: 50px;
$table-border-radius: 3px;
