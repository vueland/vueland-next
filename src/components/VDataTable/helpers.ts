export const toComparableStringFormat = val =>
  String(Number(val) || val.toLowerCase())

