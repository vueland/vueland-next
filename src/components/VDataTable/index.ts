import VDataTable from './VDataTable'

export { VDataTable }
export default VDataTable
