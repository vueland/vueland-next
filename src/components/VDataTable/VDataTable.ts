// Styles
import './VDataTable.scss'

// Vue API
import { h, watch, computed, defineComponent, ref } from 'vue'

// Effects
import { useColors } from '../../effects/use-colors'

// Components
import { VDataTableHeader } from './VDataTableHeader'
import { VDataTableBody } from './VDataTableBody'
import { VDataTableModal } from './VDataTableModal'
import { VDataTableFooter } from './VDataTableFooter'
import { VCheckbox } from '../VCheckbox'
import { VButton } from '../VButton'

// Helpers
import { toComparableStringFormat } from './helpers'

// Types
import { VNode } from 'vue'
import { Column, TableFilter } from '../../types'

export const VDataTable = defineComponent({
  name: 'v-data-table',
  props: {
    cols: {
      type: Array,
      default: () => [],
    },
    rows: {
      type: Array,
      default: () => [],
    },
    headerColor: String,
    rowsOnTable: {
      type: Array,
      default: () => [10, 15, 20, 25],
    },
    dark: Boolean,
    numbered: Boolean,
    checkbox: Boolean,
    stateOut: Boolean,
    align: String,
    color: {
      type: String,
      default: 'white',
    },
  } as any,

  setup(props, { slots, emit }) {
    const cols = ref<Column[]>([])
    const rows = ref<any[]>([])
    const checkedRows = ref<any[]>([])
    const rowsPerPage = ref<number>(20)
    const page = ref<number>(1)
    const isAllRowsChecked = ref<boolean>(false)
    const settings = ref({
      cols: false,
    })

    const filters = {}

    const { setBackground } = useColors()

    const classes = computed<Record<string, boolean>>(() => ({
      'v-data-table': true,
    }))

    const pages = computed<number>(() => {
      return Math.ceil(rows.value?.length / rowsPerPage.value)
    })

    watch(
      () => props.cols,
      to => (cols.value = to),
      { immediate: true },
    )

    watch(
      () => props.rows,
      to => (rows.value = to),
      { immediate: true },
    )

    function onCheckAll(value: boolean) {
      isAllRowsChecked.value = value
      rows.value.forEach(row => (row.checked = value))
    }

    function onCheck<T>(rows: T[]) {
      checkedRows.value = rows
      emit('checked', checkedRows.value)
    }

    function onPrevTable(num) {
      page.value = page.value > 1 ? page.value + num : page.value
    }

    function onNextTable(num: number) {
      if (rows.value.length - page.value * rowsPerPage.value > 0) {
        page.value += num
      }
    }

    function onSort(col: Column) {
      if (col.sorted) {
        col.sorted = !col.sorted
        rows.value!.reverse()
      } else {
        cols.value!.forEach(c => {
          c.sorted = col.key === c.key
        })

        sortColumn(col)
      }
    }

    function onFilter({ value, col }: TableFilter) {
      if (!value && filters[col.key]) delete filters[col.key]

      if (value) filters[col.key] = value

      if (!props.stateOut) {
        if (!Object.keys(filters).length) {
          return (rows.value = props.rows)
        }
        rows.value = filterRows(props.rows)
      } else {
        emit('filter', filters)
      }

      page.value = 1
    }

    function onSelectRowsCount(count: number) {
      rowsPerPage.value = count
    }

    function colsSettings(val: boolean, col: Column) {
      col.show = val
      emit('cols-settings', cols.value)
    }

    function sortColumn(col: Column) {
      rows.value!.sort((a, b) => {
        if (col.format) {
          if (col.format(a) > col.format(b)) return 1
        } else {
          if (a[col.key] > b[col.key]) return 1
        }
        return -1
      })
    }

    function filterRows<T>(rows: T[]) {
      const filterKeys = Object.keys(filters)

      return rows.reduce((acc, row) => {
        const rowResults: any[] = []

        filterKeys.forEach(key => {
          const { format } = cols.value!.find(col => col.key === key) as Column
          const value = format ? format(row) : row[key]

          const rowKeyValue = toComparableStringFormat(value)
          const filterValue = toComparableStringFormat(filters[key])

          if (rowKeyValue.includes(filterValue)) {
            rowResults.push(!!row[key])
          }
        })

        if (
          rowResults.length === filterKeys.length &&
          rowResults.every(value => !!value)
        ) {
          acc.push(row)
        }

        return acc
      }, [] as T[])
    }

    function genTableTools(): VNode {
      return h('div', {
          class: {
            'v-data-table__toolbar': true,
          },
        },
        {
          default: () => slots.toolbar && slots.toolbar(),
        },
      )
    }

    function genTableHeader(): VNode {
      return h(VDataTableHeader, {
        cols: cols.value,
        color: props.headerColor || props.color,
        checkbox: props.checkbox,
        dark: props.dark,
        align: props.align,
        numbered: props.numbered,
        onFilter,
        onSort,
        onCheckAll,
      })
    }

    function genTableBody(): VNode {
      return h(
        VDataTableBody,
        {
          cols: cols.value,
          rows: rows.value,
          page: page.value,
          rowsPerPage: rowsPerPage.value,
          checkbox: props.checkbox,
          checkAllRows: isAllRowsChecked.value,
          align: props.align,
          dark: props.dark,
          color: props.color,
          numbered: props.numbered,
          onCheck,
        },

        props.cols.reduce((acc, col) => {
          const slotContent = row => {
            const scoped: any = { row }

            if (col.format) {
              scoped.format = col.format
            }

            return slots[col.key] && slots[col.key]!(scoped)
          }

          if (slots[col.key]) acc[col.key] = slotContent

          return acc
        }, {}),
      )
    }

    function genTableFooter(): VNode {
      return h(VDataTableFooter, {
        pages: pages.value,
        page: page.value,
        counts: props.rowsOnTable,
        tableRowsCount: rows.value?.length,
        allRowsCount: props.rows.length,
        rowsPerPage: rowsPerPage.value,
        dark: props.dark,
        color: props.color,
        toolbar: props.toolbar,
        onPrev: onPrevTable,
        onNext: onNextTable,
        onSelect: onSelectRowsCount,
        onLastPage: val => emit('last-page', val),
        onResetPage: val => (page.value += val),
        onColsSettings: () => (settings.value.cols = true),
      })
    }

    function genTableInner(): VNode {
      return h(
        'div',
        {
          class: {
            'v-data-table__inner': true,
          },
        },
        [genTableHeader(), genTableBody()],
      )
    }

    function genColsSettingsCheckboxes(): VNode[] {
      return cols.value.map(col => h(VCheckbox, {
        label: col.title,
        modelValue: col.show,
        dark: props.dark,
        color: props.dark ? 'white' : '',
        style: { margin: '10px 0' },
        onChecked: val => colsSettings(val, col),
      }))
    }

    function genColsSettingsActions(): VNode {
      return h(VButton, {
        dark: props.dark,
        color: props.dark ? 'white' : 'primary',
        outlined: props.dark,
        label: 'ok',
        onClick: () => (settings.value.cols = false),
      })
    }

    function genColsSettingsModal(): VNode {
      return h(
        VDataTableModal,
        {
          dark: props.dark,
          color: props.color,
          show: settings.value.cols,
        },
        {
          title: () => 'Cols Settings',
          content: () => genColsSettingsCheckboxes(),
          actions: () => genColsSettingsActions(),
        },
      )
    }

    return () => {
      const propsData = {
        class: classes.value,
      }

      return h('div', setBackground(props.color, propsData), [
        slots.toolbar && genTableTools(),
        genTableInner(),
        genTableFooter(),
        genColsSettingsModal(),
      ])
    }
  },
})
