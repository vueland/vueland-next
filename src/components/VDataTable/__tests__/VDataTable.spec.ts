import { mount } from '@vue/test-utils'
import { VDataTable } from '../VDataTable'
import { cols, rows } from './mocks'
import 'regenerator-runtime/runtime'

describe('VDataTable', () => {
  let mountFunction: (options?: any) => any

  beforeEach(() => {
    mountFunction = (options = {}) => mount(VDataTable, { ...options })
  })

  it('should mound component and match snapshot', () => {
    const cmp = mountFunction()

    expect(cmp.html()).toMatchSnapshot()
  })

  it('should set cols and match snapshot', () => {
    const cmp = mountFunction({
      props: { cols },
    })

    expect(cmp.find('.v-data-table-col').exists()).toBe(true)
    expect(cmp.get('.v-data-table-col__title').text()).toBe('Name')
    expect(cmp.html()).toMatchSnapshot()
  })

  it('should test column sort and match snapshot', async () => {
    cols[0].sortable = true

    const cmp = mountFunction({
      props: { cols, rows },
    })

    expect(cmp.find('.v-data-table-col__actions-sort').exists()).toBe(true)
    expect(cmp.find('.v-data-table__row').text()).toContain('Ben')

    await cmp.find('.v-data-table-col__actions-sort').trigger('click')

    expect(cmp.find('.v-data-table-col__actions-sort--active').exists()).toBe(
      true,
    )
    expect(cmp.find('.v-data-table__row').text()).toContain('Alex')

    expect(cmp.html()).toMatchSnapshot()
  })

  it('should test column filter and match snapshot', async () => {
    cols[0].filterable = true

    const cmp = mountFunction({
      props: { cols, rows, stateOut: true },
    })

    const filterBtn = cmp.find('.v-data-table-col__actions-filter')

    expect(filterBtn.exists()).toBe(true)

    await filterBtn.trigger('click')
    const filter = cmp.find('.v-data-table-col__filter')
    const textField = cmp.find('.v-text-field__input')

    expect(textField.exists()).toBe(true)

    // await textField.trigger('input', 'Al')
    //
    // expect(cmp.emitted().filter).toHaveLength(1)
    expect(filter.exists()).toBe(true)
    expect(filter.isVisible()).toBe(true)
    // expect(cmp.find('.v-data-table__row').text()).toContain('Alex')
    expect(cmp.html()).toMatchSnapshot()
  })
})
