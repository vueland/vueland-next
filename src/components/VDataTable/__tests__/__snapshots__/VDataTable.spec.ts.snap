// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VDataTable should mound component and match snapshot 1`] = `
<div class="v-data-table white">
  <div class="v-data-table__inner">
    <div class="v-data-table__header white">
    </div>
    <div class="v-data-table__body white">
    </div>
  </div>
  <div class="v-data-table__footer">
    <div class="v-data-table__pagination">
      <div class="v-data-table__pagination-select">
        <span class="v-data-table__pagination-label">
          Rows per page
        </span>
        <div class="v-input">
          <div class="v-input__select-slot primary--text">
            <label class="v-label v-label--active v-label--on-field v-label--has-state primary--text"
                   style="left: 0px; position: absolute;"
            >
            </label>
            <div class="v-select v-validatable">
              <input class="v-select__input">
              <transition-stub>
                <div class="v-select-list elevation-4 white"
                     style="display: none;"
                >
                  <div class="v-list v-select--items-list">
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        10
                      </span>
                    </div>
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        15
                      </span>
                    </div>
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        20
                      </span>
                    </div>
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        25
                      </span>
                    </div>
                  </div>
                </div>
              </transition-stub>
            </div>
          </div>
          <div class="v-input__status">
            <transition-stub>
            </transition-stub>
          </div>
        </div>
      </div>
      <div class="v-data-table__pagination-pages">
        -
      </div>
      <div class="v-data-table__pagination-route">
        <button class="v-button elevation-3 primary"
                style="min-width: 42px;"
        >
          <i class="v-icon fas fa-chevron-left"
             style="font-size: 18px;"
          >
          </i>
        </button>
        <button class="v-button elevation-3 blue lighten-1"
                style="min-width: 42px; margin: 0px 10px;"
        >
          1
        </button>
        <button class="v-button elevation-3 primary"
                style="min-width: 42px;"
        >
          <i class="v-icon fas fa-chevron-right"
             style="font-size: 18px;"
          >
          </i>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`VDataTable should set cols and match snapshot 1`] = `
<div class="v-data-table white">
  <div class="v-data-table__inner">
    <div class="v-data-table__header white">
      <div class="v-data-table__cell text-align--start white v-data-table-col"
           style="width: 125px;"
      >
        <div class="v-data-table__cell-content">
          <div class="v-data-table-col__title">
            Name
          </div>
          <span class="v-data-table-col__actions">
          </span>
        </div>
      </div>
      <div class="v-data-table__cell text-align--start white v-data-table-col"
           style="width: 125px;"
      >
        <div class="v-data-table__cell-content">
          <div class="v-data-table-col__title">
            Email
          </div>
          <span class="v-data-table-col__actions">
          </span>
        </div>
      </div>
      <div class="v-data-table__cell text-align--start white v-data-table-col"
           style="width: 125px;"
      >
        <div class="v-data-table__cell-content">
          <div class="v-data-table-col__title">
            Body
          </div>
          <span class="v-data-table-col__actions">
          </span>
        </div>
      </div>
    </div>
    <div class="v-data-table__body white">
    </div>
  </div>
  <div class="v-data-table__footer">
    <div class="v-data-table__pagination">
      <div class="v-data-table__pagination-select">
        <span class="v-data-table__pagination-label">
          Rows per page
        </span>
        <div class="v-input">
          <div class="v-input__select-slot primary--text">
            <label class="v-label v-label--active v-label--on-field v-label--has-state primary--text"
                   style="left: 0px; position: absolute;"
            >
            </label>
            <div class="v-select v-validatable">
              <input class="v-select__input">
              <transition-stub>
                <div class="v-select-list elevation-4 white"
                     style="display: none;"
                >
                  <div class="v-list v-select--items-list">
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        10
                      </span>
                    </div>
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        15
                      </span>
                    </div>
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        20
                      </span>
                    </div>
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        25
                      </span>
                    </div>
                  </div>
                </div>
              </transition-stub>
            </div>
          </div>
          <div class="v-input__status">
            <transition-stub>
            </transition-stub>
          </div>
        </div>
      </div>
      <div class="v-data-table__pagination-pages">
        -
      </div>
      <div class="v-data-table__pagination-route">
        <button class="v-button elevation-3 primary"
                style="min-width: 42px;"
        >
          <i class="v-icon fas fa-chevron-left"
             style="font-size: 18px;"
          >
          </i>
        </button>
        <button class="v-button elevation-3 blue lighten-1"
                style="min-width: 42px; margin: 0px 10px;"
        >
          1
        </button>
        <button class="v-button elevation-3 primary"
                style="min-width: 42px;"
        >
          <i class="v-icon fas fa-chevron-right"
             style="font-size: 18px;"
          >
          </i>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`VDataTable should test column filter and match snapshot 1`] = `
<div class="v-data-table white">
  <div class="v-data-table__inner">
    <div class="v-data-table__header white">
      <div class="v-data-table__cell text-align--start white v-data-table-col v-data-table-col--sorted"
           style="width: 125px;"
      >
        <div class="v-data-table__cell-content">
          <div class="v-data-table-col__title">
            Name
          </div>
          <span class="v-data-table-col__actions">
            <button class="v-icon v-icon--link v-icon--clickable fas fa-arrow-up v-data-table-col__actions-sort v-data-table-col__actions-sort--active"
                    style="font-size: 14px;"
            >
            </button>
            <button class="v-icon v-icon--link v-icon--clickable fas fa-align-justify v-data-table-col__actions-filter"
                    style="font-size: 14px;"
            >
            </button>
          </span>
          <div class="v-data-table-col__filter white"
               style
          >
            <span class="v-data-table-col__filter-header">
              Name
            </span>
            <div class="v-input">
              <div class="v-input__field-slot primary--text">
                <div class="v-input__prepend-icon">
                  <i class="v-icon fas fa-search primary--text"
                     style="font-size: 16px;"
                     dark="false"
                  >
                  </i>
                </div>
                <div class="v-input__clear">
                  <button class="v-icon v-icon--link v-icon--clickable fas fa-times primary--text"
                          style="font-size: 16px;"
                          dark="false"
                  >
                  </button>
                </div>
                <label class="v-label v-label--on-field primary--text"
                       style="left: 0px; position: absolute;"
                >
                  insert
                </label>
                <div class="v-text-field v-validatable">
                  <input class="v-text-field__input"
                         type="text"
                  >
                </div>
              </div>
              <div class="v-input__status">
                <transition-stub>
                </transition-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-data-table__cell text-align--start white v-data-table-col"
           style="width: 125px;"
      >
        <div class="v-data-table__cell-content">
          <div class="v-data-table-col__title">
            Email
          </div>
          <span class="v-data-table-col__actions">
          </span>
        </div>
      </div>
      <div class="v-data-table__cell text-align--start white v-data-table-col"
           style="width: 125px;"
      >
        <div class="v-data-table__cell-content">
          <div class="v-data-table-col__title">
            Body
          </div>
          <span class="v-data-table-col__actions">
          </span>
        </div>
      </div>
    </div>
    <div class="v-data-table__body white">
      <div class="v-data-table__row">
        <div class="v-data-table__cell text-align--start undefined"
             style="width: 125px;"
        >
          <div class="v-data-table__cell-content">
            Alex
          </div>
        </div>
        <div class="v-data-table__cell text-align--start undefined"
             style="width: 125px;"
        >
          <div class="v-data-table__cell-content">
            alex@mail.ru
          </div>
        </div>
        <div class="v-data-table__cell text-align--start undefined"
             style="width: 125px;"
        >
          <div class="v-data-table__cell-content">
            some body text
          </div>
        </div>
      </div>
      <div class="v-data-table__row">
        <div class="v-data-table__cell text-align--start undefined"
             style="width: 125px;"
        >
          <div class="v-data-table__cell-content">
            Ben
          </div>
        </div>
        <div class="v-data-table__cell text-align--start undefined"
             style="width: 125px;"
        >
          <div class="v-data-table__cell-content">
            ben@mail.ru
          </div>
        </div>
        <div class="v-data-table__cell text-align--start undefined"
             style="width: 125px;"
        >
          <div class="v-data-table__cell-content">
            some body text
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-data-table__footer">
    <div class="v-data-table__pagination">
      <div class="v-data-table__pagination-select">
        <span class="v-data-table__pagination-label">
          Rows per page
        </span>
        <div class="v-input">
          <div class="v-input__select-slot primary--text">
            <label class="v-label v-label--active v-label--on-field v-label--has-state primary--text"
                   style="left: 0px; position: absolute;"
            >
            </label>
            <div class="v-select v-validatable">
              <input class="v-select__input">
              <transition-stub>
                <div class="v-select-list elevation-4 white"
                     style="display: none;"
                >
                  <div class="v-list v-select--items-list">
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        10
                      </span>
                    </div>
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        15
                      </span>
                    </div>
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        20
                      </span>
                    </div>
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        25
                      </span>
                    </div>
                  </div>
                </div>
              </transition-stub>
            </div>
          </div>
          <div class="v-input__status">
            <transition-stub>
            </transition-stub>
          </div>
        </div>
      </div>
      <div class="v-data-table__pagination-pages">
        1 - 2 from 2
      </div>
      <div class="v-data-table__pagination-route">
        <button class="v-button elevation-3 primary"
                style="min-width: 42px;"
        >
          <i class="v-icon fas fa-chevron-left"
             style="font-size: 18px;"
          >
          </i>
        </button>
        <button class="v-button elevation-3 blue lighten-1"
                style="min-width: 42px; margin: 0px 10px;"
        >
          1
        </button>
        <button class="v-button elevation-3 primary"
                style="min-width: 42px;"
        >
          <i class="v-icon fas fa-chevron-right"
             style="font-size: 18px;"
          >
          </i>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`VDataTable should test column sort and match snapshot 1`] = `
<div class="v-data-table white">
  <div class="v-data-table__inner">
    <div class="v-data-table__header white">
      <div class="v-data-table__cell text-align--start white v-data-table-col v-data-table-col--sorted"
           style="width: 125px;"
      >
        <div class="v-data-table__cell-content">
          <div class="v-data-table-col__title">
            Name
          </div>
          <span class="v-data-table-col__actions">
            <button class="v-icon v-icon--link v-icon--clickable fas fa-arrow-up v-data-table-col__actions-sort v-data-table-col__actions-sort--active"
                    style="font-size: 14px;"
            >
            </button>
          </span>
        </div>
      </div>
      <div class="v-data-table__cell text-align--start white v-data-table-col"
           style="width: 125px;"
      >
        <div class="v-data-table__cell-content">
          <div class="v-data-table-col__title">
            Email
          </div>
          <span class="v-data-table-col__actions">
          </span>
        </div>
      </div>
      <div class="v-data-table__cell text-align--start white v-data-table-col"
           style="width: 125px;"
      >
        <div class="v-data-table__cell-content">
          <div class="v-data-table-col__title">
            Body
          </div>
          <span class="v-data-table-col__actions">
          </span>
        </div>
      </div>
    </div>
    <div class="v-data-table__body white">
      <div class="v-data-table__row">
        <div class="v-data-table__cell text-align--start undefined"
             style="width: 125px;"
        >
          <div class="v-data-table__cell-content">
            Alex
          </div>
        </div>
        <div class="v-data-table__cell text-align--start undefined"
             style="width: 125px;"
        >
          <div class="v-data-table__cell-content">
            alex@mail.ru
          </div>
        </div>
        <div class="v-data-table__cell text-align--start undefined"
             style="width: 125px;"
        >
          <div class="v-data-table__cell-content">
            some body text
          </div>
        </div>
      </div>
      <div class="v-data-table__row">
        <div class="v-data-table__cell text-align--start undefined"
             style="width: 125px;"
        >
          <div class="v-data-table__cell-content">
            Ben
          </div>
        </div>
        <div class="v-data-table__cell text-align--start undefined"
             style="width: 125px;"
        >
          <div class="v-data-table__cell-content">
            ben@mail.ru
          </div>
        </div>
        <div class="v-data-table__cell text-align--start undefined"
             style="width: 125px;"
        >
          <div class="v-data-table__cell-content">
            some body text
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-data-table__footer">
    <div class="v-data-table__pagination">
      <div class="v-data-table__pagination-select">
        <span class="v-data-table__pagination-label">
          Rows per page
        </span>
        <div class="v-input">
          <div class="v-input__select-slot primary--text">
            <label class="v-label v-label--active v-label--on-field v-label--has-state primary--text"
                   style="left: 0px; position: absolute;"
            >
            </label>
            <div class="v-select v-validatable">
              <input class="v-select__input">
              <transition-stub>
                <div class="v-select-list elevation-4 white"
                     style="display: none;"
                >
                  <div class="v-list v-select--items-list">
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        10
                      </span>
                    </div>
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        15
                      </span>
                    </div>
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        20
                      </span>
                    </div>
                    <div class="v-list-item">
                      <span class="v-list-item__title">
                        25
                      </span>
                    </div>
                  </div>
                </div>
              </transition-stub>
            </div>
          </div>
          <div class="v-input__status">
            <transition-stub>
            </transition-stub>
          </div>
        </div>
      </div>
      <div class="v-data-table__pagination-pages">
        1 - 2 from 2
      </div>
      <div class="v-data-table__pagination-route">
        <button class="v-button elevation-3 primary"
                style="min-width: 42px;"
        >
          <i class="v-icon fas fa-chevron-left"
             style="font-size: 18px;"
          >
          </i>
        </button>
        <button class="v-button elevation-3 blue lighten-1"
                style="min-width: 42px; margin: 0px 10px;"
        >
          1
        </button>
        <button class="v-button elevation-3 primary"
                style="min-width: 42px;"
        >
          <i class="v-icon fas fa-chevron-right"
             style="font-size: 18px;"
          >
          </i>
        </button>
      </div>
    </div>
  </div>
</div>
`;
